src/v/archival/ntp_archiver_service.cc

```diff
@@ -1581,18 +1522,87 @@ ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
         .max_timestamp = upload.max_timestamp,
         .delta_offset = delta,
         .ntp_revision = _rev,
-        .archiver_term = upload_ctx.archiver_term,
+        .archiver_term = archiver_term,
         .segment_term = upload.term,
         .delta_offset_end = delta_offset_next,
         .sname_format = cloud_storage::segment_name_format::v3,
         .metadata_size_hint = tx_size,
       },
       .name = upload.exposed_name, .delta = offset - base,
       .stop = ss::stop_iteration::no,
-      .upload_kind = upload_ctx.upload_kind,
+      .upload_kind = upload_kind,
     };
 }
 
+ss::future<ntp_archiver::scheduled_upload>
+ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
+    auto start_upload_offset = upload_ctx.start_offset;
+    auto last_stable_offset = upload_ctx.last_offset;
+
+    auto log = _parent.log();
+
+    candidate_creation_result candidate_result;
+
+    switch (upload_ctx.upload_kind) {
+    case segment_upload_kind::non_compacted:
+        candidate_result = co_await _policy.get_next_candidate(
+          start_upload_offset,
+          last_stable_offset,
+          log,
+          *_parent.get_offset_translator_state(),
+          _conf->segment_upload_timeout);
+        break;
+    case segment_upload_kind::compacted:
+        const auto& m = manifest();
+        candidate_result = co_await _policy.get_next_compacted_segment(
+          start_upload_offset, log, m, _conf->segment_upload_timeout);
+        break;
+    }
+
+    co_return co_await ss::visit(
+      candidate_result,
+      [](std::monostate) -> ss::future<scheduled_upload> {
+          vassert(false, "Unexpected default upload candidate creation result");
+      },
+      [this, upload_ctx](upload_candidate_with_locks& candidate_with_locks) {
+          return maybe_schedule_single_upload(
+            std::move(candidate_with_locks),
+            upload_ctx.archiver_term,
+            upload_ctx.upload_kind);
+      },
+      [this, upload_ctx](skip_offset_range& skip_offsets) {
+          const auto log_level = log_level_for_error(skip_offsets.reason);
+          vlogl(
+            _rtclog,
+            log_level,
+            "Failed to make upload candidate, skipping offset range: {}",
+            skip_offsets);
+          return ss::make_ready_future<scheduled_upload>(scheduled_upload{
+            .result = std::nullopt,
+            .inclusive_last_offset = skip_offsets.end_offset,
+            .meta = std::nullopt,
+            .name = std::nullopt,
+            .delta = std::nullopt,
+            .stop = ss::stop_iteration::yes,
+            .upload_kind = upload_ctx.upload_kind,
+          });
```

Comment:
Are we allowing a gap in the uploaded log in this case?

Reply:
The idea was to have the upload logic remain unchanged, and just make the code paths explicit. So this branch corresponds to the earlier:

```
    if (
      upload.sources.empty()
      && upload_ctx.upload_kind == segment_upload_kind::compacted
      && model::offset{} != upload.final_offset) {
        vlog(
          _rtclog.warn,
          "Upload skipped for range: {}-{} because these offsets lie inside "
          "batches or the reupload is not smaller than the current segment",
          upload.starting_offset,
          upload.final_offset);
        co_return scheduled_upload{
          .result = std::nullopt,
          .inclusive_last_offset = upload.final_offset,
          .meta = std::nullopt,
          .name = std::nullopt,
          .delta = std::nullopt,
          .stop = ss::stop_iteration::yes,
          .upload_kind = upload_ctx.upload_kind,
        };
    }
```

But now I see that there is nothing guaranteeing that this `skip_offset_range` is only used for reupload. Currently this is true, because this is only returned in cases where:

* offset is inside batch in segment reupload
* upload candidate size is unchanged from original in segment reupload

but there is no constraint that someone cannot return this in the wrong place (eg during normal upload path) and create a gap in the log in cloud. So this looks like a potential bug. 

Maybe we can check here if upload kind is compacted, then return the final offset, otherwise `{}`.

Reply:
Thinking more about this, conditionally returning different offsets here is risky. It might be better to assert that if the result type is `skip_offset_range` then the upload kind is `compacted`.